proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,305 @@
+```
+
+#### Downstream Clusters
+
+Conceptually the _downstream cluster_ represents the cluster that the kafka client connects to.  The downstream clusters list specifies
+all the kafka clustsers are being presented by Kroxylicious. 
+
+A _downstream clustser_ references exactly one _upstream cluster_. 
+
+A _downstream cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the upstream cluster.
+
+A _downstream cluster_ enumerates the downstream brokers that comprise the downstream cluster.
+
+##### Downstream Brokers
+
+Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
+for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.
+
+Each downstream broker references exactly one upstream broker which must exist within the referenced upstream cluster.  It is an error if there is not a
+one-to-one correspondence between downstream and upstream brokers.
+
+TLS key material may be provided either at the downstream cluster level or at the individual broker level.
+
+
+```yaml
+- name: my-public-cluster
+  upstreamClusterRef: my-private-cluster
+  filterChainRef: my-upstream-filter-chain
+  brokers:
+  - name: broker-1
+    endpointRef: my-tls-endpoint
+    sniMatchAddress: broker-1-public.example.com
+    upstreamBrokerRef: broker-1
+    # or brokers themselves provide the key/cert
+    - tls:
+        key:
+        cert:
+  - name: broker-2
+    ...
+  # cluster wide TLS cert
+  tls:
+    key:
+    cert:
+```
+
+
+#### Upstream Clusters
+
+An _upstream cluster_ represents a physical kafka cluster.   The upstream clusters list specifies all the kafka clustsers that are being exposed through Kroxylicious.  
+
+An _upstream cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers
+of the physical cluster.  
+
+An _upstream cluster_ enumerates the upstream brokers that comprise the physical cluster.
+
+#### Upstream Brokers
+
+Each upstream broker specifies the address of the physical broker and whether TLS is to be used.  It will also specify trust settings.
+
+TLS trust material may be provided at the upstream broker level or at the individual broker level.
+
+```
+- name: my-private-cluster
+  filterChainRef: my-downstream-filter-chain
+  tls:
+    trust:
+    ...
+  brokers:
+  - name: broker-1
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+      ...
+  - name: broker-2
+    address: broker-1-private:9092
+    ...
+    tls:
+      trust:
+```
+
+#### Filter Chain
+
+Currently kroxylicious provide a single filter chain.  It will refactored so that we can support many filter chains, each identified by name. The `downstream cluster` and `upstream cluster` may reference a filter chain. 
```

Comment:
I'm starting to think it might be better to express this as the `virtual` cluster defines a filter chain (potentially applied by the `frontendHandler` and the `physical` cluster defines a filter chain (again applied by the `backendHandler`). Those `virtual` and `physical` chain definitions could potentially be shared between multiples of each cluster type but there is no single over arching chain definition.

Reply:
Yes, maybe.  if each virtual cluster/cluster declares the filter chain in its own right, there's no problem passing cluster specific configuration.